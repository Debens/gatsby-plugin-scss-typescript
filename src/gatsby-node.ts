// eslint-disable-next-line @typescript-eslint/tslint/config
import 'core-js/stable';

import { RuleSetUseItem } from 'webpack';

type ObjectRule = Exclude<RuleSetUseItem, string | Function>;

export const onCreateWebpackConfig = ({ stage, plugins, rules, actions }, options) => {
    const { setWebpackConfig } = actions;
    const isProduction = !stage.includes('develop');

    const {
        cssMinifyOptions,
        cssExtractOptions,
        cssLoaderOptions,
        sassLoaderOptions,
        declarationOptions,
    } = options;

    const typeLoader: ObjectRule = {
        loader: '@teamsupercell/typings-for-css-modules-loader',
        options: declarationOptions,
    };

    const useCss: ObjectRule[] = rules.css(cssLoaderOptions).use;
    const useCssModules: ObjectRule[] = rules.cssModules(cssLoaderOptions).use;

    const sassLoaders: ObjectRule[] = [
        { loader: 'resolve-url-loader' },
        {
            loader: 'sass-loader',
            options: {
                sourceMap: !isProduction,
                ...sassLoaderOptions,
            },
        },
    ];

    const isCssLoader = ({ loader }: ObjectRule) =>
        !!['/css-loader/', '\\css-loader\\'].find(matcher => loader.includes(matcher));

    /**
     * https://github.com/TeamSupercell/typings-for-css-modules-loader/blob/master/src/index.js#L64-L66
     *
     * TeamSupercell use the locals property in the named exports.
     *
     * Sadly this means we have to disable the tree shaking when generating the typings to allow the typings generation.
     * This wil have no effect on the actual tree-shaking of course
     */
    const toDisabledTreeShaking = loader => ({
        ...loader,
        options: {
            ...loader.options,
            modules: {
                ...loader.options.modules,
                namedExport: false,
            },
        },
    });

    const css = useCss.concat(sassLoaders);
    const cssModules = useCssModules
        .flatMap(loader =>
            isCssLoader(loader) ? [typeLoader, toDisabledTreeShaking(loader)] : loader,
        )
        .concat(sassLoaders);

    switch (stage) {
        case 'develop':
        case 'build-javascript':
        case 'build-html':
        case 'develop-html': {
            setWebpackConfig({
                module: {
                    rules: [
                        {
                            oneOf: [
                                {
                                    test: /\.module\.s(a|c)ss$/,
                                    use: cssModules,
                                },
                                {
                                    test: /\.s(a|c)ss$/,
                                    use: css,
                                },
                            ],
                        },
                    ],
                },
                optimization: {
                    minimizer: [plugins.minifyCss(cssMinifyOptions)],
                },
                plugins: [
                    plugins.extractText(cssExtractOptions),
                    plugins.ignore(/css\.d\.ts$/),
                ],
            });
        }
    }
};
