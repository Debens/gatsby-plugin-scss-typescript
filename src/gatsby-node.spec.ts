import { onCreateWebpackConfig } from './gatsby-node';

const rules = {
    css: args => ({
        test: /\.css$/,
        use: [
            { loader: require.resolve('style-loader') },
            { loader: require.resolve('css-loader'), options: args },
            { loader: require.resolve('sass-loader') },
        ],
    }),
    cssModules: args => ({
        test: /\.module\.css$/,
        use: [
            { loader: require.resolve('style-loader') },
						{ loader: require.resolve('dts-css-modules-loader') },
            {
                loader: require.resolve('css-loader'),
                options: { ...args, modules: true },
            },
            { loader: require.resolve('sass-loader') },	
        ],
    }),
};

const plugins = {
    extractText: args => `extractText(${JSON.stringify(args)})`,
    ignore: regex => `ignore(${regex})`,
    minifyCss: args => `minifyCss(${JSON.stringify(args)})`,
};

const actions = { setWebpackConfig: jest.fn() };

describe('gatsby-plugin-sass', () => {
    beforeEach(() => {
        actions.setWebpackConfig.mockReset();
    });

    const stages = ['develop', 'build-javascript', 'develop-html', 'build-html'];

    stages.forEach(stage => {
        describe(`given stage ${stage}`, () => {
            describe('when creating webpack config', () => {
                describe('and passing no options', () => {
                    it('then should configure the config', () => {
                        onCreateWebpackConfig({ actions, stage, rules, plugins }, {});
                        expect(actions.setWebpackConfig).toMatchSnapshot();
                    });
                });

                describe('sass loader options', () => {
                    const sassLoaderOptions = {
                        includePaths: ['absolute/path/a', 'absolute/path/b'],
                    };

                    it('then should configure the config', () => {
                        onCreateWebpackConfig(
                            { actions, stage, rules, plugins },
                            { sassLoaderOptions },
                        );
                        expect(actions.setWebpackConfig).toMatchSnapshot();
                    });
                });

                describe('css loader options', () => {
                    const cssLoaderOptions = {
                        importLoaders: 1,
                        localIdentName: '[name]_[local]___[hash:base64:5]_[emoji:1]',
                    };

                    it('then should configure the config', () => {
                        onCreateWebpackConfig(
                            { actions, stage, rules, plugins },
                            { cssLoaderOptions },
                        );
                        expect(actions.setWebpackConfig).toMatchSnapshot();
                    });
                });
            });
        });
    });
});
