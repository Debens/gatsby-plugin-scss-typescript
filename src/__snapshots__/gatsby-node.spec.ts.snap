// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-sass given stage build-html when creating webpack config and passing no options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage build-html when creating webpack config css loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage build-html when creating webpack config sass loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage build-javascript when creating webpack config and passing no options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage build-javascript when creating webpack config css loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage build-javascript when creating webpack config sass loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": false,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": false,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop when creating webpack config and passing no options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop when creating webpack config css loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop when creating webpack config sass loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop-html when creating webpack config and passing no options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop-html when creating webpack config css loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "importLoaders": 1,
                        "localIdentName": "[name]_[local]___[hash:base64:5]_[emoji:1]",
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-sass given stage develop-html when creating webpack config sass loader options then should configure the config 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "@teamsupercell/typings-for-css-modules-loader",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": Object {
                        "modules": Object {
                          "namedExport": false,
                        },
                      },
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "/node_modules/style-loader/index.js",
                    },
                    Object {
                      "loader": "/node_modules/css-loader/dist/cjs.js",
                      "options": undefined,
                    },
                    Object {
                      "loader": "/node_modules/postcss-loader/src/index.js",
                    },
                    Object {
                      "loader": "resolve-url-loader",
                    },
                    Object {
                      "loader": "sass-loader",
                      "options": Object {
                        "includePaths": Array [
                          "absolute/path/a",
                          "absolute/path/b",
                        ],
                        "sourceMap": true,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        "optimization": Object {
          "minimizer": Array [
            "minifyCss(undefined)",
          ],
        },
        "plugins": Array [
          "extractText(undefined)",
          "ignore(/css\\\\.d\\\\.ts$/)",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
